boskos/client/client.go

```diff
@@ -450,43 +471,62 @@ func (c *Client) Release(name, dest string) error {
 	values.Set("name", name)
 	values.Set("dest", dest)
 	values.Set("owner", c.owner)
-	resp, err := c.httpPost("/release", values, "", nil)
-	if err != nil {
-		return err
-	}
-	defer resp.Body.Close()
 
-	if resp.StatusCode != http.StatusOK {
-		return fmt.Errorf("status %s, statusCode %v releasing %s", resp.Status, resp.StatusCode, name)
+	work := func(retriedErrs *[]error) (bool, error) {
+		resp, err := c.httpPost("/release", values, "", nil)
+		if err != nil {
+			*retriedErrs = append(*retriedErrs, err)
+			return false, nil
+		}
+		defer resp.Body.Close()
+
+		if resp.StatusCode != http.StatusOK {
+			*retriedErrs = append(*retriedErrs, fmt.Errorf("status %s, statusCode %v releasing %s", resp.Status, resp.StatusCode, name))
+			return false, nil
+		}
+		return true, nil
 	}
-	return nil
+
+	return retry(work)
 }
 
 // Update a resource on the server, setting the state and user data
 func (c *Client) Update(name, state string, userData *common.UserData) error {
-	var body io.Reader
+	var bodyData *bytes.Buffer
 	if userData != nil {
-		b := new(bytes.Buffer)
-		err := json.NewEncoder(b).Encode(userData)
+		bodyData = new(bytes.Buffer)
+		err := json.NewEncoder(bodyData).Encode(userData)
 		if err != nil {
 			return err
 		}
-		body = b
 	}
 	values := url.Values{}
 	values.Set("name", name)
 	values.Set("owner", c.owner)
 	values.Set("state", state)
-	resp, err := c.httpPost("/update", values, "application/json", body)
-	if err != nil {
-		return err
-	}
-	defer resp.Body.Close()
 
-	if resp.StatusCode != http.StatusOK {
-		return fmt.Errorf("status %s, status code %v updating %s", resp.Status, resp.StatusCode, name)
+	work := func(retriedErrs *[]error) (bool, error) {
+		// As the body is an io.Reader and hence its content
+		// can only be read once, we have to copy it for every request we make
+		var body io.Reader
+		if bodyData != nil {
+			body = bytes.NewBuffer(bodyData.Bytes())
+		}
```

Comment:
A `bytes.Reader` can be used instead.

Reply:
Not on line 511, otherwise we will do a NPD down the line because of "interface is only nil if type and value are unset". Updated to a Reader in line 513